Complete Django start documentation (#2468)
Scalingo/documentation: src/_posts/languages/python/django/2000-01-01-start.md

tags: python django tutorial getting-started-tutorial
index: 1
---

{% include info_tutorial_requirements.md %}

## Create Your Application and Databases on Scalingo

{% note %}
  You can also use our web dashboard to achieve this operation
{% endnote %}

```bash
$ scalingo create myapp
```

* Go on the [dashboard](https://dashboard.scalingo.com/apps) of your application.
* Select the __Addons__ category
* Choose the database you want to use PostgreSQL®, MySQL®...

## Initialize Your Application locally

On your computer, start a Django website.

```bash
# Create your project
mkdir my-project
cd my-project

# use pipenv to handle dependancies (recommanded for python buildpack)
pip install --user pipenv

# Install Django and other useful packages:
pipenv install django django-environ psycopg gunicorn dj-static

# enter the virtual environment
pipenv shell

# Initialize Django project, this command will
# create a manage.py file and a myapp directory
django-admin startproject myapp .
```

{% note %}
  If you've selected MySQL database instead of PostreSQL,
  do not install psycopg but mysqlclient instead.

  `pipenv install django mysqlclient gunicorn ...`
{% endnote %}

{% note %}

>   Pipenv is the prefered way to manage dependencies. It will had Pipfile and Pipfile.lock
>   files, therefor deployment system will consider your application as a Python app.
> 
>   [Please refere to Python buildpack for more information on this
>   subject]({% post_url languages/python/2000-01-01-start %})

Reviewer: issue: a few typos + we prefer using present tense in our documentation.
```suggestion
  Pipenv is the preferred way to manage dependencies. It adds Pipfile and Pipfile.lock
  files, therefore deployment system considers your application as a Python app.

  [Please refer to Python buildpack for more information on this
  subject]({% post_url languages/python/2000-01-01-start %})
```

{% endnote %}

## Define How To Start Your Application

You need to specify how to start the applicative server, and define it to display logs
on `stdout` to fit the [12-factor](https://12factor.net/) principles.

> This is acheived by create a file named `Procfile` at the root of the project with a web entry.

Reviewer: issue: typos

```suggestion
This is achieved by creating a file named `Procfile` at the root of the project with a web entry.
```

> Scalingo handle WSGI and ASGI server :

Reviewer: ```suggestion
Scalingo handles WSGI and ASGI server :
```

```yaml
web: gunicorn myapp.wsgi --log-file -
```

```yaml
web: gunicorn myapp.asgi --worker-class=uvicorn.workers.UvicornWorker --log-file -
```

{% note %}
  For more details, check the documentation about [Procfile]({% post_url platform/app/2000-01-01-procfile %}).
{% endnote %}

### Start script

> A better practice is to use a start script, usually stored in bin directory. This
> will give you more flexibility on how to start your app.

Reviewer: nitpick(non-blocking): I would say that the directory name should be formatted as code.

```suggestion
A better practice is to use a start script, usually stored in the `bin` directory. This
gives you more flexibility on how to start your app.
```

scalingo --app app-name env-set ALLOWED_HOSTS=app-name.osc-fr1.scalingo.io,example.com
```

## Deploy your application

Your app is almost done, now we need to ship it. Django requires that severals
operation are done each time we deploy the app. In this chapter we will see how
we can do that on Scalingo.

> In Python buildpack, there is 3 hooks we can use: `pre_compile`, `post_compil`
> and `post_deploy`, each is triggered at a specific time and will have some use.

Reviewer: issue: IMO hooks are not code but feature names. I would also keep a link to the documentation

```suggestion
In Python buildpack, there are 3 hooks we can use: pre-compile, post-compile
and [postdeploy]({% post_url platform/app/2000-01-01-postdeploy-hook %}). 
Each is triggered at a specific time and will have some use.
```

### Automatic Database Migrations

> If you use Django's ORM and migrations files, you want to use `post_deploy` hook

Reviewer: nitpick(non-blocking): 
```suggestion
If you use Django's ORM and migrations files, you want to use postdeploy hook
```

The `migrate` command is the one to execute to apply to migrations to your
database. You can add it to your `Procfile` like this:

First create a file `bin/post_deploy.sh` with:

```bash
#!/bin/sh

# Execute structure migrations
python manage.py migrate users
python manage.py migrate
```

After, activate the hook by adding following line in `Procfile`:

```bash
postdeploy: bash bin/post_deploy.sh
```

> This will trigger a one-off container just after finishing compiling

Reviewer: ```suggestion
This triggers a one-off container just after finishing compiling
```

but before replacing current app (if any were running when triggering
the deployment).

{% note %}
  More information in [postdeploy hook section]({% post_url platform/app/2000-01-01-postdeploy-hook %})
{% endnote %}

### Collect static

Previous hook is not good for collect static because the script is

> run in a one-off which is a copy of the final container. Therefor all

Reviewer: nitpick (non-blocking): from what I read, the syntax without the "e" is more formal. I find it excessive in the context of this documentation.

```suggestion
run in a one-off which is a copy of the final container. Therefore all
```

files manipulation would be available only on the one-off and not
in the actual container which will serve web request.

To trigger script on the actual container, you will need to use

> `post_compile` hook.

Reviewer: ```suggestion
post-compile hook.
```

Add a file `bin/post_compile` with (note the missing .sh extension!):

```bash
#!/bin/sh

export PYTHONPATH=/build/${REQUEST_ID}/.apt/usr/lib/python3/dist-packages/:${PYTHONPATH}

# collect static
python manage.py collectstatic --noinput
```

### Compile `.mo` Message Translation File

If you are using `compilemessages` command to generate `.mo` file for gettext
translations, that command needs to run at the same stage as collect static.

Add following line to `bin/post_compile` file:

```bash
python manage.py compilemessages
```

> ## Save Your App With Git and automatic deployment

Reviewer: ```suggestion
## Save Your App With Git and Automatic Deployment
```

scalingo --app myapp git-show
```

> Each time you will push on master branch, it should trigger a deployment job on Scalingo.

Reviewer: ```suggestion
Each time you push on the `master` branch, it should trigger a deployment job on Scalingo.
```

{% note %}
  Please see [Continuous Integration section]({% post_url platform/deployment/2000-01-01-deploy-with-git %}) for more information on git.
{% endnote %}

## Access Your Application

> The url is displayed at the end of the deployment logs. It won't change.

Reviewer: ```suggestion
The URL is displayed at the end of the deployment logs. It won't change.
```
